Added metadata fields under Settings tab

utkarshcmu opened the pull request:
Changes in this PR:
1. Create/Update dashboard http api has userId parameter now, which is completely optional to maintain backward compatibility.
2. Under "Manage dashboard" > "Settings" > "Metadata", now "Created by:" details are displayed too.

Fixing 1 part out of 3 from #3841

utkarshcmu committed:
Added createdBy in metadata ui and dashboard table

utkarshcmu committed:
Api stores dashboard creator

utkarshcmu committed:
Updated http_api docs

utkarshcmu committed:
Added dtos and UI for more metadata

torkelo commented on docs/sources/reference/http_api.md:
```diff
@@ -70,13 +70,15 @@ Creates a new dashboard or updates an existing dashboard.
       },
-      "overwrite": false
+      "overwrite": false,
+      "userId:": 3
```
why is this in the write request? The UserId is taken from request user/auth context.

utkarshcmu replied:
I got your point, will remove this. :)

torkelo commented on pkg/models/dashboards.go:
```diff
@@ -91,8 +92,11 @@ func NewDashboardFromJson(data map[string]interface{}) *Dashboard {
 // GetDashboardModel turns the command into the savable model
 func (cmd *SaveDashboardCommand) GetDashboardModel() *Dashboard {
 	dash := NewDashboardFromJson(cmd.Dashboard)
+	if dash.Data["version"] == 0 {
```
should probably use dash.Version here :)

utkarshcmu replied:
Actually, we should not use dash.Version here because dash.Version is not set properly until this line of code. Anyway, I tried using dash.Version as suggested by you but then saving dashboard is updating "created_by" field too in the DB while saving the dashboard again, which is incorrect. Also, If you look at line 83 in the same file, you will find that we use the same if condition to create the dashboard object. So, I think lets stick with this if condition and not change it for the correct functionality. :)

torkelo commented:
Great PR :heart: , just some minor tweaks and it go in :)

utkarshcmu committed:
Added more metadata

utkarshcmu commented:
This PR shows this metadata in the "Settings" panel:

<img width="802" alt="screen shot 2016-01-28 at 9 53 54 am" src="https://cloud.githubusercontent.com/assets/1019609/12653596/1b8f8f3c-c5a5-11e5-9f13-57118fb2f3a7.png">

utkarshcmu commented:
Please review/merge. :)

bergquist commented:
Wouldn't it be easier/better to get data from the dashboard object in the frontend? 
ping @torkelo

utkarshcmu commented:
@bergquist

I thought about getting the data from frontend too,
But then two reasons came into my mind to get the data from backend:

1. Why to put load on frontend to calculate these stats.
2. When half of the metadata comes from the backend, then why not get everything in the same dashboardMeta object to keep everything stndardized.

torkelo commented:
I dont think the dashboard stats belong in the dashboard meta object

utkarshcmu commented:
All the fields like created, updated and updated by were already coming in dashboardMeta object, so I made it a standard for other fields also come in dashboardMeta object and it works well as the calulations are done in the backend. 

Another thought would be to get it as a separate object as dashboardStats. But this requires more work. May be we can start with this PR as it does not hurt any performance. Later with time we can move these fields to dashboard stats. What do you guys think?

utkarshcmu commented:
We can even leverage a new api as 
/api/dashboards/db/db-slug/stats or similar to get the data in a completely new fashion. Just a thought. :)

utkarshcmu commented:
I think its a good starting point. We can make it better like all other features with time.

utkarshcmu commented:
Any updates on this?

torkelo commented:
Sorry, only one minor thing. there are still some total stats on the dashboard meta object that do not belong there

utkarshcmu commented:
OK. So, you want me to remove "Total rows, panels, queries" from this PR? Does "Version" also needs to be removed?

woodsaj commented:
The "total*" stats are computed values. They should just be computed in the frontend when rendering the settings tab.

utkarshcmu commented:
@woodsaj Got your point. Will make the changes asap.

utkarshcmu commented:
@bergquist / @torkelo All changes made as suggested and resubmitted in #3918. Please review/merge #3918. After merging, we can close this PR. 

Lost track of this repo, that is why created another PR. :(

utkarshcmu commented:
Closing due to #3918
